x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,155 @@
+
+package mssql
+
+import (
+	"database/sql"
+
+	"github.com/pkg/errors"
+
+	s "github.com/elastic/beats/libbeat/common/schema"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	// Register driver.
+	_ "github.com/denisenkom/go-mssqldb"
+)
+
+// NewFetcher is called from every metricset to initialize their fetching routines. It opens a different connection to
+// the database for each fetch operation
+func NewFetcher(uri string, qs []string, schema *s.Schema, log *logp.Logger) (*Fetcher, error) {
+	db, err := sql.Open("sqlserver", uri)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create db instance")
+	}
+
+	// Check the connection before executing all queries to reduce the number
+	// of connection errors that we might encounter.
+	if err = db.Ping(); err != nil {
+		return nil, err
+	}
+
+	f := &Fetcher{
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries sequentially to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string // List of queries to execute concurrently.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
+		}
+	}
+}
+
+type rowsResultHandler struct {
+	dest        []interface{}
+	rows        *sql.Rows
+	reporter    mb.ReporterV2
+	columnNames []string
+}
+
+func newRowsResult(reporter mb.ReporterV2, columnNames []string, rows *sql.Rows) *rowsResultHandler {
+	return &rowsResultHandler{
+		dest:        make([]interface{}, len(columnNames)),
+		columnNames: columnNames,
+		rows:        rows,
+		reporter:    reporter,
+	}
+}
+
+// getEventsWithQuery performs the query on the database and creates a rowsResultHandler to handle them
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) (err error) {
+	var rows *sql.Rows
+	rows, err = db.Query(query)
+	if err != nil {
+		return errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer func() {
+		if err2 := rows.Close(); err != nil {
+			if err != nil {
+				err = errors.Wrap(err, err2.Error())
+			} else {
+				err = err2
+			}
+		}
+	}()
```

Comment:
Sorry, but as you said this is actually a bit ugly :grimacing: Also [according to documentation](https://golang.org/pkg/database/sql/#Rows.Close) rows are closed automatically by Next, so this should only be needed if the Scan/Next loop is unexpectedly interrupted (having rows creation and the Scan/Next loop in the same place would be another reason in favour of [this proposal](https://github.com/elastic/beats/pull/9414#discussion_r241210948)).

Also for better or worse it is quite usual to ignore the error on calls to `Close()`, even in the examples in the documentation for the `sql` module... I think we'd be fine with reducing this to `defer Close()`, or at much log this error at the debug level, because an error here is probably very very rare.

I'd change it here, but we will have to revisit this code in any case, so as you prefer... but if you keep this here, please at least use a more meaningful name for the inner error, like `closeErr`.

Reply:
I think we should not block this PR with more stylish and subjective things.

A very very rare errors that is hidden on purpose leads to a very very difficult to debug, discover and fix bugs so also very very difficult to maintain. That's way I prefer explicit approaches (say ugly if you want) Than just to obfuscate the error.

So please, if you don't mind, let's move this PR forward because it has been hangling for months already, it has 81 comments and there's still has a lot of work to do with the upcoming metricsets that will need a lot of review in many aspects.

Reply:
I think the problem is that code that gets merge gets rarely changed again. My main issue with the above is that when I read the code, I don't really understand what it's doing above.

To keep this moving the following suggestion: Please add it to the module meta issue as an entry that has to be looked at again so we don't forget about the cleanup.

Reply:
I'm happy with any proposal that doesn't involve a function with 4 or 5 responsibilities :1st_place_medal: In any case, I already have some notes for the next PR that I'm planning to write in the header of the upcoming metricset PR once it's open. I already have to change few stylish things also in this particular file so we have time to discuss a way that is _easy for everybody_ to understand / look :smile: